clang/runtime/dpct-rt/include/dpl_extras/memory.h

```diff
@@ -917,34 +928,56 @@ device_pointer<T> malloc_device(const std::size_t num_elements) {
 static inline device_pointer<void> malloc_device(const std::size_t num_bytes) {
   return device_pointer<void>(num_bytes);
 }
-template <typename T>
-device_pointer<T> device_new(device_pointer<T> p, const T &value,
+
+#ifdef DPCT_USM_LEVEL_NONE
+template <typename T, typename U>
+device_pointer<T> device_new(device_pointer<U> p, const T &value,
+                             const std::size_t count) {
+  auto converted_buf = p.buffer.template reinterpret<T>(sycl::range<1>(count));
+  ::std::uninitialized_fill(
+      oneapi::dpl::execution::make_device_policy(dpct::get_default_queue()),
+      oneapi::dpl::begin(converted_buf), oneapi::dpl::end(converted_buf),
+      value);
+  return device_pointer<T>(converted_buf, p.idx);
+}
+// buffer manages lifetime
+template <typename T> void free_device(device_pointer<T> ptr) {}
```

Comment:
Should put in namespace "internal"

Reply:
`free_device` is actually a public function that is used to free `device_malloc` calls and is not an internal utility.

Reply:
> `free_device` is a public function that is used to free `device_malloc` calls and is not an internal utility.

Reply:
Why name free_device instead of device_free?  The name stype is different from device device_new, device_delete.

Reply:
I believe this is just modifying code which has existed here for some time, not new APIs (so the naming decision was not made in this PR).  However, if `free_device` is not already in use by SYCLomatic, I would also be in favor of switching it to `device_free` as well.

Reply:
free_device has been used in the migrated code https://github.com/oneapi-src/SYCLomatic/blob/1b1335850c4b7fcde30659dcfef08bffbd340e94/clang/test/dpct/thrust-cast.cu#L37

So, I'm OK to current solution.